crates/turbopack-css/src/process.rs

```diff
@@ -902,3 +937,134 @@ impl Issue for ParsingIssue {
+        stylesheet::{ParserOptions, StyleSheet},
+        visitor::Visit,
+    };
+    use swc_core::{
+        common::{FileName, FilePathMapping},
+        css::{ast::Stylesheet, parser::parser::ParserConfig, visit::VisitWith},
+    };
+
+    use super::{CssError, CssValidator};
+
+    fn lint_lightningcss(code: &str) -> Vec<CssError> {
+        let mut ss = StyleSheet::parse(
+            code,
+            ParserOptions {
+                css_modules: Some(lightningcss::css_modules::Config {
+                    pattern: Pattern::default(),
+                    dashed_idents: false,
+                }),
+                ..Default::default()
+            },
+        )
+        .unwrap();
+
+        let mut validator = CssValidator { errors: Vec::new() };
+        ss.visit(&mut validator).unwrap();
+
+        validator.errors
+    }
+
+    fn lint_swc(code: &str) -> Vec<CssError> {
+        let cm = swc_core::common::SourceMap::new(FilePathMapping::empty());
+
+        let fm = cm.new_source_file(FileName::Custom("test.css".to_string()), code.to_string());
+
+        let ss: Stylesheet = swc_core::css::parser::parse_file(
+            &fm,
+            None,
+            ParserConfig {
+                css_modules: true,
+                ..Default::default()
+            },
+            &mut vec![],
+        )
+        .unwrap();
+
+        let mut validator = CssValidator { errors: Vec::new() };
+        ss.visit_with(&mut validator);
+
+        validator.errors
+    }
+
+    #[track_caller]
+    fn assert_lint_success(code: &str) {
+        assert_eq!(lint_lightningcss(code), vec![]);
+        assert_eq!(lint_swc(code), vec![]);
+    }
+
+    #[track_caller]
+    fn assert_lint_failure(code: &str) {
+        assert_ne!(lint_lightningcss(code), vec![]);
+        assert_ne!(lint_swc(code), vec![]);
+    }
+
+    #[test]
+    fn css_module_pure_lint() {
+        assert_lint_success(
+            "html {
+            --foo: 1;
+        }",
+        );
+
+        assert_lint_success(
+            "#id {
+            color: red;
+        }",
+        );
+
+        assert_lint_success(
+            ".class {
+            color: red;
+        }",
+        );
+
+        assert_lint_success(
+            "html.class {
+            color: red;
+        }",
+        );
+
+        assert_lint_failure(
+            "div {
+            color: red;
+        }",
+        );
+
+        assert_lint_failure(
+            "div > span {
+            color: red;
+        }",
+        );
+
+        assert_lint_failure(
+            "div span {
+            color: red;
+        }",
+        );
+
+        assert_lint_failure(
+            "div[data-foo] {
+            color: red;
+        }",
+        );
+
+        assert_lint_failure(
+            "div[data-foo=\"bar\"] {
+            color: red;
+        }",
+        );
+
+        assert_lint_failure(
+            "div[data-foo=\"bar\"] span {
+            color: red;
+        }",
+        );
```

Comment:
- `*`  ❌
- `[data-foo]` ❌
- `:not(.class)` ❌
- `.class:not(div)` ✅
- `:not(div)` ❌